### PR TITLE
Make derefer work everwhere

### PR DESCRIPTION
**Pull request by ouz-a**
Follow up work on previous PR's #95649 and #95857.

r? rust-lang/mir-opt

_Co-Authored-By: @oli-obk_

### CODE REVIEW

File: compiler/rustc_mir_transform/src/deref_separator.rs [hunk 1]
```diff
@@ -1,68 +1,89 @@
...
+
+impl<'tcx> MutVisitor<'tcx> for DerefChecker<'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_place(&mut self, place: &mut Place<'tcx>, _: PlaceContext, loc: Location) {
+        let mut place_local = place.local;
+        let mut last_len = 0;
+        let mut last_deref_idx = 0;
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/deref_separator.rs:23]: I think we can probably figure out a way to write the algorithm without needing this variable, but seems ok for now

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_mir_transform/src/deref_separator.rs (hunk 1 above)
Comment (ouz-a) [reply compiler/rustc_mir_transform/src/deref_separator.rs:23]: I tried to find a solution that would work for every case but there are a lot of cases...

File: compiler/rustc_middle/src/mir/patch.rs [hunk 2]
```diff
@@ -143,8 +142,7 @@ impl<'tcx> MirPatch<'tcx> {
         let mut delta = 0;
         let mut last_bb = START_BLOCK;
         let mut terminator_targets = Vec::new();
-        let mut inf_and_stmt: VecDeque<(SourceInfo, StatementKind<'_>)> = VecDeque::new();
-
+        let mut statements:Vec<Statement<'_>> = Vec::new();
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/patch.rs]: Please merge this vec with the `terminator_targets`

Comment (ouz-a) [reply compiler/rustc_middle/src/mir/patch.rs]: Sorry I totally understood a different thing 😅

File: compiler/rustc_middle/src/mir/patch.rs [hunk 3]
```diff
@@ -172,9 +171,8 @@ impl<'tcx> MirPatch<'tcx> {
             delta += 1;
         }
 
-        for target in terminator_targets.iter().rev() {
-            let stmt = statements.pop().unwrap();
-            body[*target].statements.insert(0, stmt);
+        for (stmt, target) in stmts_and_targets.iter().rev() {
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/patch.rs:174]: If you use `into_iter` you can avoid the clone inside

**Comment by oli-obk** [comment]
r? @oli-obk 

Reviewing even though I co-authored as my contribution was a mir opt test

@bors try @rust-timer queue

**Comment by oli-obk** [comment]
Const eval now has more statements to chew through, so this perf change is expected. 

@bors r+

**Comment by JakobDegen** [comment]
I was talking about a related issue on discord, and wanted to mention this here. This code:
```rust
let x = 0;
let y = &x;
let p = unsafe { core::mem::transmute::<&&i32,&&mut i32>(&y) };
let _r = &**p;
```
Currently passes Miri. That will continue to be the case after this change, but only because this runs after retag statements are added (and does not add them itself). We now have the option to make the above code report UB. It is unclear to me if this is something we want. cc @RalfJung and rust-lang/unsafe-code-guidelines#270 .

**Comment by ouz-a** [comment]
>  but only because this runs after retag statements are added

The next goal is to pull deref_separator early as possible

**Comment by JakobDegen** [comment]
Ah, ok, apparently my comment was less off-topic than I thought it was :sweat_smile: . In that case, this will make the code shared in the other thread:
```rust
let x = 0;
let y = &x;
let p = unsafe { core::mem::transmute::<&&i32,&&mut i32>(&y) };
let _r = &**p;
```
Report as UB in Miri. cc @RalfJung (sorry about the ghost ping before), is this something we want to prevent?

**Comment by RalfJung** [comment]
> The next goal is to pull deref_separator early as possible, so this is going to run before retag

I don't think we want that code to be UB though. I haven't quite understood yet why this pass would have the stated effect.

**Comment by oli-obk** [comment]
It shouldn't have, if it does, we'll deploy the appropriate fixes immediately instead of when we move this to before borrowck

**Comment by RalfJung** [comment]
However, I should note that conceptually, giving the temporary for `*p` the type `&mut T` is also quite right -- we don't actually own it at that type there. The type `&T` would be more accurate.

Basically, we have to ensure that all passes that run after this just don't care about the type of a reference stored in a local -- `&` and `&mut` must make no difference, since this pass makes types broken. That's why moving this before `Retag` would be an issue, but I imagine this could also be a problem for other passes so I think this should be approached more systematically.

**Comment by oli-obk** [comment]
> we have to ensure that all passes that run after this just don't care about the type of a reference stored in a local -- `&` and `&mut` must make no difference, since this pass makes types broken.

It does at present. Before actually moving it earlier we were considering various strategies. While we could start messing with types and using different ones, that seems suboptimal. What we came up with in the compiler brainstorm meetings during rust all hands was more of a "mark these locals in a way that analysis passes can understand". The types are correct as far as I can tell, it's just that the locals are more temporary than regular temporaries.

The extreme alternative would be to replace all such locals' types with raw pointers which will make these analyses ignore them mostly. We'll still need markers for borrowck though, even in that case

**Comment by RalfJung** [comment]
`&mut` is not a correct type, at least not under the notions of "correct" I would usually apply to MIR.

For example, if we were to obtain such MIR as input to a formal methods tool like Prusti or Creusot, we'd definitely *not* want to try to treat this as a mutable reference. That would fail spectacularly.

**Comment by oli-obk** [comment]
@bors retry

Makes sense, we'll move this pass before retag and emit the correct types before doing any further actions then.

**Comment by oli-obk** [comment]
All of the primary regressions and the tt-muncher secondary regressions are similar to

```
 20,293,864  ???:<rustc_expand::mbe::macro_parser::TtParser>::parse_tt
 17,155,452  ???:<alloc::rc::Rc<alloc::vec::Vec<(rustc_ast::tokenstream::TokenTree, rustc_ast::tokenstream::Spacing)>> as core::ops::drop::Drop>::drop
-12,949,133  ???:<alloc::vec::Vec<(rustc_ast::tokenstream::TokenTree, rustc_ast::tokenstream::Spacing)> as core::ops::drop::Drop>::drop
```

This looks to me like just some LLVM inliner noise. We do have more local variables now, which LLVM previously just had as intermediate LLVM Values that were rarely (if ever) put into a local. This seems to me like a problem that we could address in ways beyond fixing this PR's regression. After all, if the user writes out the temporary local variables manually instead of us expanding their code, then they would see the same issues.

The other major regression in a secondary benchmark (the const eval stress test) is expected, the const evaluator has to process more statements, and every statement has overhead. Similar to the LLVM case, there are avenues to explore here to make const eval faster. I believe there are still some low hanging fruit that will yield measurable improvements, but we need to dig into benchmark dumps to find out where exactly they lie.

@rustbot label: +perf-regression-triaged

**Pull request merged**